pkg/interop/client.go

```diff
@@ -0,0 +1,136 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interop
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+const (
+	// loRaAllianceDomain is the domain of LoRa Alliance.
+	loRaAllianceDomain = "lora-alliance.org"
+
+	// loRaAllianceJoinEUIDomain is the LoRa Alliance domain used for JoinEUI resolution.
+	loRaAllianceJoinEUIDomain = "joineuis." + loRaAllianceDomain
+
+	// loRaAllianceNetIDDomain is the LoRa Alliance domain used for NetID resolution.
+	loRaAllianceNetIDDomain = "netids." + loRaAllianceDomain
+)
+
+func LoRaAllianceJoinServerFQDN(eui types.EUI64) string {
+	o := []rune(strings.ToLower(eui.String()))
+	r := make([]rune, 2*len(o)-1)
+	// Inspired by https://github.com/golang/example/blob/46695d81d1fae905a270fb7db8a4d11a334562fe/stringutil/reverse.go#L20-L27
+	for i, j := 0, len(o)-1; i < len(o)/2; i, j = i+1, j-1 {
+		r[i+i], r[j+j] = o[j], o[i]
+		r[i+i+1] = '.'
+		r[j+j-1] = '.'
+	}
+	return fmt.Sprintf("%s.%s", string(r), loRaAllianceJoinEUIDomain)
+}
+
+func HandleJoinRequest(ctx context.Context, tlsConfig *tls.Config, netID types.NetID, joinEUI, devEUI types.EUI64, req *ttnpb.JoinRequest, fqdn string) (*ttnpb.JoinResponse, error) {
+	dlSettings, err := lorawan.MarshalDLSettings(req.DownlinkSettings)
+	if err != nil {
+		return nil, ErrJoinReq.WithCause(err)
+	}
+
+	var cfList []byte
+	if req.CFList != nil {
+		cfList, err = lorawan.MarshalCFList(*req.CFList)
+		if err != nil {
+			return nil, ErrJoinReq.WithCause(err)
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	if err := json.NewEncoder(buf).Encode(&JoinReq{
+		NsJsMessageHeader: NsJsMessageHeader{
+			MessageHeader: MessageHeader{
+				ProtocolVersion: "1.0",
+				MessageType:     MessageTypeJoinReq,
+				// TODO: Add TransactionID if necessary
+				// TODO: Add SenderToken if necessary
+			},
+			SenderID:   NetID(netID),
```

Comment:
@johanstokking should the NS's `NetID` be filled in both `SenderdID` and `SenderNSID`?
What are these for? Is it for distinguishing sNS/fNS?

Reply:
Yes exactly